openvasd/openvasd.c

```diff
@@ -0,0 +1,2251 @@
+    {
+    case POST:
+      if (data != NULL && data[0] != '\0')
+        {
+          // SET Content type
+          customheader =
+            curl_slist_append (customheader, "Content-Type: application/json");
+          // Set body
+          curl_easy_setopt (curl, CURLOPT_POSTFIELDS, data);
+          curl_easy_setopt (curl, CURLOPT_POSTFIELDSIZE, strlen (data));
+        }
+      break;
+    case GET:
+      curl_easy_setopt (curl, CURLOPT_HTTPGET, 1L);
+      break;
+    case DELETE:
+      curl_easy_setopt (curl, CURLOPT_CUSTOMREQUEST, "DELETE");
+      break;
+    default:
+      curl_easy_setopt (curl, CURLOPT_CUSTOMREQUEST, "HEAD");
+      break;
+    };
+
+  if (customheader != NULL)
+    curl_easy_setopt (curl, CURLOPT_HTTPHEADER, customheader);
+
+  // Init the struct where the response is stored and set the callback function
+  curl_easy_setopt (curl, CURLOPT_WRITEFUNCTION, response_callback_fn);
+  curl_easy_setopt (curl, CURLOPT_WRITEDATA, resp);
+
+  return curl;
+}
+
+/** @brief Send request
+ *
+ *  @param[in] curl        The CURL handler to perform an request.
+ *  @param[in] header_name If this field is set, is looked in the header and
+ *                         its value is returned inside the response.
+ *  @param[out] response   The response struct to be filled with the response
+ * code and the header value.
+ *
+ *  @return Return struct containing the http response code and the response
+ * body. In case of error the struct is filled with code RESP_CODE_ERR (-1) and
+ * a message. NULL on memory related error. Response must be free()'ed by the
+ * caller with openvasd_response_free()
+ */
+static openvasd_resp_t
+openvasd_send_request (CURL *curl, const char *header_name,
+                       openvasd_resp_t response)
+{
+  long http_code = RESP_CODE_ERR;
+
+  int ret = CURLE_OK;
+  if ((ret = curl_easy_perform (curl)) != CURLE_OK)
+    {
+      g_warning ("%s: Error sending request: %d", __func__, ret);
+      curl_easy_cleanup (curl);
+      curl = NULL;
+      response->code = http_code;
+      response->body = g_strdup ("{\"error\": \"Error sending request\"}");
+      return response;
+    }
+
+  curl_easy_getinfo (curl, CURLINFO_RESPONSE_CODE, &http_code);
+
+  if (header_name && *header_name)
+    {
+      struct curl_header *hname;
+      curl_easy_header (curl, header_name, 0, CURLH_HEADER, -1, &hname);
+      response->header = g_strdup (hname->value);
+    }
+  curl_easy_cleanup (curl);
+  curl = NULL;
+  response->code = http_code;
+
+  return response;
+}
+
+/**
+ * @brief Request HEAD
+ *
+ * @param conn Connector struct with the data necessary for the connection
+ *
+ * @return Response containing the header information
+ */
+openvasd_resp_t
+openvasd_get_version (openvasd_connector_t *conn)
+{
+  char *err = NULL;
+  CURL *hnd = NULL;
+  openvasd_resp_t response = NULL;
+  openvasd_stringstream resp;
+
+  response = g_malloc0 (sizeof (struct openvasd_response));
+  if (response == NULL)
```

Comment:
I see these checks are everywhere. I'd really like them removed because glib always aborts when out of mem.